rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,449 @@
+
+## Non-goals
+It is a non-goal for this project to provide an application that is expected to be a part of Linkerd application architecture (control / data plane). This project in no way shall directly affect the way Linkerd functions. Unlike the existing integration tests, the e2e tests for conformance in this project are not expected to be a part of the Linkerd CI workflow either. These tests shall run as a stand-alone component such that it can interact with a k8s cluster.
+
+## Deliverables
+
+### Must have (shall be completed by end of GSoC):
+- An e2e test suite that can perform conformance validation for the following features:
+    1. Automatic proxy injection
+    2. Linkerd tap, stat, routes, edges cmd
+    3. Distributed tracing
+    4. Canary release
+    5. Ingress
+    6. gRPC streaming (client/server/bi-directional streaming and unary)
+    7. Websockets
+    8. MySQL
+    9. Redis
+- A sample emojivoto app with feature flags that can enable/disable features required for conformance vaidation.
+
+### Good-to-have (if extra time remains, these shall be worked on):
+- The conformance validation testing suite shall also cover the following features :
+    1. Telemetry and metrics
+    2. Getting route based metrics (Service profiles, retries, timeouts, fault injection)
+    3.Checking if data plane proxies return 503
+    4. mTLS
+    5. Cluster security and introspection
+    6. Data plane PSP
+    7. Elastic search
+- A new sample application - MovieChat - that has all the features required from conformance validation perspective
+
+
+# Prior art
+
+[prior-art]: #prior-art
+- Sample sonobuoy plugins - https://github.com/vmware-tanzu/sonobuoy/tree/master/examples/plugins
+- K8s e2e tests - https://github.com/kubernetes/kubernetes/tree/master/test/e2e
+- K8s sonobuoy image - https://github.com/kubernetes/kubernetes/tree/master/cluster/images/conformance
+- Some sample applications developed by linkerd for testing:
+    - https://github.com/BuoyantIO/emojivoto
+    - https://github.com/BuoyantIO/booksapp
+    - https://github.com/linkerd/linkerd-examples
+    - https://github.com/BuoyantIO/slow_cooker/
+
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+This project proposes that tests be written in golang mainly due to how easy it is to manage outputs and assertions, organize tests and maintain readability. Having the option to use client-go k8s API could be another plus point. However, shell script may be a good tool too to write these tests, but we should also consider factors such as contributor experience. Would love to have the community’s opinion regarding this.
+
+
+# Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+As a Linkerd user, it is important to check if Linkerd works as expected even outside of the emojivoto / any other sample application. Setting the workloadNamespace property in the test configuration file gives the users this flexibility :
```

Comment:
Try with "As a Linkerd user, it is important to know that Linkerd works with my services, not just emojivoto".

Backtick "workloadNamespace"